File: impl/src/org/ufomatics/archipelago/view/swing/FieldPane.java
package org.ufomatics.archipelago.view.swing;

import org.ufomatics.archipelago.control.UFOPlatformController;
import org.ufomatics.archipelago.model.MetaField;
import org.ufomatics.archipelago.model.MetaType;

import javax.swing.*;

public class FieldPane extends AbstractPane {
    /**
     * The data. Will be modified if ok, or not if cancelled
     */
    private MetaField data;

    private JTextField nameTextfield;
    private JComboBox typeCombo;
    private JTextArea descriptionTextarea;

    private JButton newTypeButton;

    private UFOPlatformController controller;

    public FieldPane(UFOPlatformController controller, MetaField field) {
        this.controller = controller;
        this.data = field;

        SpringLayout layout = new SpringLayout();
        setLayout(layout);

        JLabel nameLabel = new JLabel("Name");
        nameTextfield = new JTextField(field.getName());
        JLabel typeLabel = new JLabel("Type");
        typeCombo = createTypeCombo(controller);
        MetaType selectedType = field.getType() == null ? controller.getDefaultFieldType() : field.getType();
        typeCombo.setSelectedItem(selectedType);
        add(nameLabel);
        add(nameTextfield);
        add(typeLabel);
        add(typeCombo);
        layout.putConstraint(SpringLayout.EAST, this, 5, SpringLayout.EAST, typeCombo);
        layout.putConstraint(SpringLayout.SOUTH, this, 5, SpringLayout.SOUTH, typeCombo);
        layout.putConstraint(SpringLayout.EAST, nameTextfield, 5, SpringLayout.EAST, typeCombo);
        layout.putConstraint(SpringLayout.WEST, nameLabel, 5, SpringLayout.WEST, this);
        layout.putConstraint(SpringLayout.NORTH, nameLabel, 5, SpringLayout.NORTH, this);
        layout.putConstraint(SpringLayout.WEST, nameTextfield, 5, SpringLayout.EAST, nameLabel);
        layout.putConstraint(SpringLayout.NORTH, nameTextfield, 0, SpringLayout.NORTH, nameLabel);
        layout.putConstraint(SpringLayout.WEST, typeLabel, 0, SpringLayout.WEST, nameLabel);
        layout.putConstraint(SpringLayout.NORTH, typeLabel, 5, SpringLayout.SOUTH, nameLabel);
        layout.putConstraint(SpringLayout.WEST, typeCombo, 0, SpringLayout.WEST, nameTextfield);
        layout.putConstraint(SpringLayout.NORTH, typeCombo, 5, SpringLayout.SOUTH, nameTextfield);
    }

    public boolean requestFocusInWindow() {
        return nameTextfield.requestFocusInWindow();
    }

    public void ok() {
        data.setName(nameTextfield.getText());
        data.setType((MetaType) typeCombo.getSelectedItem());
    }
}

File: src/archipelago/impl/src/ufomatics/archipelago/view/swing/FieldPane.ts
import {AbstractPane} from "archipelago/impl/src/ufomatics/archipelago/view/swing/AbstractPane"
import {SpringLayout, SpringLayoutDirection} from "archipelago/api/util/jsdk/swing/SpringLayout"
import {UFOPlatformController} from "archipelago/api/control/UFOPlatformController"
import {MetaField} from "archipelago/api/model/MetaField"
import {JLabel} from "archipelago/api/util/jsdk/swing/JLabel"
import {JTextField} from "archipelago/api/util/jsdk/swing/JTextField"
import {MetaType} from "archipelago/api/model/MetaType"
import {JComboBox} from "archipelago/api/util/jsdk/swing/JComboBox"

export class FieldPane extends AbstractPane {

  private readonly nameTextfield: JTextField
  private readonly typeCombo: JComboBox

  /*private descriptionTextarea: JTextArea
  private newTypeButton: JButton*/

  constructor(
    protected controller: UFOPlatformController,
    /**
     * The data. Will be modified if ok, or not if cancelled
     */
    protected data: MetaField) {
    super()
    const layout = new SpringLayout()
    this.setLayout(layout)

    const nameLabel = new JLabel("Name")
    this.nameTextfield = new JTextField(this.data.getName())
    const typeLabel = new JLabel("Type")
    this.typeCombo = this.createTypeCombo(controller)
    const selectedType = this.data.getType() == null ? controller.getDefaultFieldType() : this.data.getType()
    this.typeCombo.setSelectedItem(selectedType)
    this.add(nameLabel)
    this.add(this.nameTextfield)
    this.add(typeLabel)
    this.add(this.typeCombo)
    layout.putConstraint(SpringLayoutDirection.EAST, this, 5, SpringLayoutDirection.EAST, this.typeCombo)
    layout.putConstraint(SpringLayoutDirection.SOUTH, this, 5, SpringLayoutDirection.SOUTH, this.typeCombo)
    layout.putConstraint(SpringLayoutDirection.EAST, this.nameTextfield, 5, SpringLayoutDirection.EAST, this.typeCombo)
    layout.putConstraint(SpringLayoutDirection.WEST, nameLabel, 5, SpringLayoutDirection.WEST, this)
    layout.putConstraint(SpringLayoutDirection.NORTH, nameLabel, 5, SpringLayoutDirection.NORTH, this)
    layout.putConstraint(SpringLayoutDirection.WEST, this.nameTextfield, 5, SpringLayoutDirection.EAST, nameLabel)
    layout.putConstraint(SpringLayoutDirection.NORTH, this.nameTextfield, 0, SpringLayoutDirection.NORTH, nameLabel)
    layout.putConstraint(SpringLayoutDirection.WEST, typeLabel, 0, SpringLayoutDirection.WEST, nameLabel)
    layout.putConstraint(SpringLayoutDirection.NORTH, typeLabel, 5, SpringLayoutDirection.SOUTH, nameLabel)
    layout.putConstraint(SpringLayoutDirection.WEST, this.typeCombo, 0, SpringLayoutDirection.WEST, this.nameTextfield)
    layout.putConstraint(SpringLayoutDirection.NORTH, this.typeCombo, 5, SpringLayoutDirection.SOUTH, this.nameTextfield)
  }

  requestFocusInWindow(): boolean {
    return this.nameTextfield.requestFocusInWindow()
  }

  ok(): void {
    this.data.setName(this.nameTextfield.getText())
    this.data.setType(this.typeCombo.getSelectedItem() as MetaType)
  }
}
